src/NLog/Internal/TargetWithFilterChain.cs

```diff
@@ -50,6 +51,115 @@ internal class TargetWithFilterChain
         /// </summary>
         private StackTraceUsage? _stackTraceUsage;
 
+        struct CallSiteKey : IEquatable<CallSiteKey>
+        {
+            public CallSiteKey(string methodName, string fileSourceName, int fileSourceLineNumber)
+            {
+                MethodName = methodName ?? string.Empty;
+                FileSourceName = fileSourceName ?? string.Empty;
+                FileSourceLineNumber = fileSourceLineNumber;
+            }
+
+            public readonly string MethodName;
+            public readonly string FileSourceName;
+            public readonly int FileSourceLineNumber;
+
+            /// <summary>
+            /// Serves as a hash function for a particular type.
+            /// </summary>
+            /// <returns>
+            /// A hash code for the current <see cref="T:System.Object"/>.
+            /// </returns>
+            public override int GetHashCode()
+            {
+                return MethodName.GetHashCode() ^ FileSourceName.GetHashCode() ^ FileSourceLineNumber;
+            }
+
+            /// <summary>
+            /// Determines if two objects are equal in value.
+            /// </summary>
+            /// <param name="obj">Other object to compare to.</param>
+            /// <returns>True if objects are equal, false otherwise.</returns>
+            public override bool Equals(object obj)
+            {
+                return obj is CallSiteKey key && Equals(key);
+            }
+
+            /// <summary>
+            /// Determines if two objects of the same type are equal in value.
+            /// </summary>
+            /// <param name="other">Other object to compare to.</param>
+            /// <returns>True if objects are equal, false otherwise.</returns>
+            public bool Equals(CallSiteKey other)
+            {
+                return FileSourceLineNumber == other.FileSourceLineNumber
+                    && string.Equals(FileSourceName, other.FileSourceName, StringComparison.Ordinal)
+                    && string.Equals(MethodName, other.MethodName, StringComparison.Ordinal);
+            }
+        }
+
+        internal bool TryCallSiteClassNameOptimization(StackTraceUsage stackTraceUsage, LogEventInfo logEvent)
+        {
+            if (stackTraceUsage != StackTraceUsage.WithCallSite
+#if !SILVERLIGHT
+             && stackTraceUsage != StackTraceUsage.WithCallSiteSource
+#endif
+               )
+                return false;
+
+            if (logEvent.CallSiteInformation == null)
```

Comment:
we prefer to combine this with below, `  if (string.IsNullOrEmpty(logEvent.CallSiteInformation?.CallerFilePath))`

Reply:
Done